include/ofi_mem.h

```diff
@@ -200,89 +200,111 @@ void dummy ## name (void) /* work-around global ; scope */
 /*
  * Buffer pool (free stack) template for shared memory regions
  */
-#define SMR_FREESTACK_EMPTY	NULL
+#define SMR_ALIGN_BOUNDARY	64
+#define SMR_FREESTACK_EMPTY	(-1)
+
+struct smr_freestack {
+	uint64_t		entry_base_offset;
+	size_t			object_size;
+	size_t			size;
+	int16_t			free;
+	int16_t			top;
+	int16_t 		entry_next[];
+};
 
-#define SMR_FREESTACK_HEADER 					\
-	void		*base_addr;				\
-	size_t		size;					\
-	void		*next;					\
+#define smr_freestack_isempty(fs)	((fs)->top == SMR_FREESTACK_EMPTY)
 
-#define smr_freestack_isempty(fs)	((fs)->next == SMR_FREESTACK_EMPTY)
-#define smr_freestack_push(fs, local_p)				\
-do {								\
-	void *p = (char **) fs->base_addr +			\
-	    ((char **) ofi_freestack_get_next(local_p) -	\
-		(char **) fs);					\
-	*(void **) ofi_freestack_get_next(local_p) = (fs)->next;\
-	(fs)->next = p;						\
-} while (0)
-#define smr_freestack_pop(fs) smr_freestack_pop_impl(fs, fs->next)
+static inline void* smr_freestack_get_entry_from_index(struct smr_freestack *fs,
+		int16_t index)
+{
+	return (void*) (((char *) fs) + fs->entry_base_offset +
+			(fs->object_size * index));
+}
 
-static inline void* smr_freestack_pop_impl(void *fs, void *next)
+static inline long freestack_size(int elem_size, int num_elements)
 {
-	void *local;
+	return (sizeof(struct smr_freestack) + sizeof(int16_t) * num_elements +
+			elem_size * num_elements + SMR_ALIGN_BOUNDARY);
+}
 
-	struct _freestack {
-		SMR_FREESTACK_HEADER
-	} *freestack = (struct _freestack*) fs;
-	assert(next != NULL);
+/* Push by entry_index */
+static inline void smr_freestack_push_by_index(struct smr_freestack *fs,
+		int16_t entry_index)
+{
+	fs->entry_next[entry_index] = fs->top;
+	fs->top = entry_index;
+	fs->free++;
+}
 
-	local = (char **) fs + ((char **) next -
-		(char **) freestack->base_addr);
+/* Push by entry_offset */
+static inline void smr_freestack_push_by_offset(struct smr_freestack *fs,
+		uint64_t entry_offset)
+{
+        smr_freestack_push_by_index(fs,
+                        (entry_offset - fs->entry_base_offset) /
+                        fs->object_size);
+}
 
-	freestack->next = *((void **)local);
-	ofi_freestack_init_next(local);
+/* Push by object */
+static inline void smr_freestack_push(struct smr_freestack *fs, void *local_p)
+{
+        smr_freestack_push_by_offset(fs,
+                ((char *) local_p - (char*) fs));
+}
 
-	return ofi_freestack_get_user_buf(local);
+#define SMR_DECLARE_FREESTACK(name) struct smr_freestack name ## _fs
```

Comment:
I think you should be able to remove this completely. No need to declare a specific type of freestack

Reply:
We will remove this